### PR TITLE
Add contributor guideline

### LINKED ISSUE #11: Document Contribution Guidelines
As an Eiffel Community Member, I would like the contribution guidelines documented so that I know how to contribute.

### PR DESCRIPTION
**Pull request by fdegir**
Closes #11

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+   limitations under the License.
+--->
+
+# Contributing to Eiffel Community
+Thank you for contributing to the Eiffel Community!
+
+The following is a set of guidelines for contributing to this repository.
+
+## How to Propose Changes
+Anyone is welcome to propose changes to the contents of this repository by creating a new [Issue](https://github.com/eiffel-community/eiffel/issues) ticket in GitHub. These requests may concern anything contained in the repo: changes to documentation, bug fixes, requests for additional information, additional event types, requests for additional examples, new featuers et cetera.
```
Comment (fredjn) [inline CONTRIBUTING.md]: Since this contributor guideline now refers to all eiffel-community projects I guess linking to the eiffel protocol project specifically might be a bit confusing.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+1. What type of change is requested?
+1. Why would you like to see this change?
+1. Can you provide any concrete examples?
+1. Which arguments in favor can you think of?
+1. Which arguments against can you think of, and why are they outweighed by the arguments in favor?
+
+Also, keep in mind that just as anyone is welcome to propose a change, anyone is welcome to disagree with and criticize that proposal.
+
+### Closing Issues
+An Issue can be closed by any member of [the repository maintainers' team](https://github.com/orgs/eiffel-community/teams/eiffel-protocol-maintainers). This can happen in various ways, for varying reasons:
```
Comment (fredjn) [inline CONTRIBUTING.md]: This link is broken, for me at least.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+Contributions can be made by anyone using the standard [GitHub Fork and Pull model](https://help.github.com/articles/about-pull-requests). When making a pull request, keep a few things in mind.
+1. Always explicitly connect a pull request to an Issue (as indiciated by the Issue template).
+1. Make sure you target the correct branch. If you are unsure which branch is appropriate, ask in the Issue thread.
+1. Pull Requests will be publicly reviewed, criticized, and potentially rejected. Don't take it personally.
+
+### Reviewing and Merging Pull Requests
+We use the Squash and Merge model, which means that all commits in a Pull Request get squashed into a single commit in the target branch. In other words, the revision history will look like a string of single commits corresponding one-to-one with Issues.
+
+Pull requests can be merged by members of the [the repository maintainers' team](https://github.com/orgs/eiffel-community/teams/eiffel-protocol-maintainers). There is a certain protocol to adhere to, however, as well as expectations on membership.
```
Comment (fredjn) [inline CONTRIBUTING.md]: Broken link

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+## Scope
+
+This Code of Conduct applies both within project spaces and in public spaces when an individual is representing the project or its community. Examples of representing a project or community include using an official project e-mail address, posting via an official social media account, or acting as an appointed representative at an online or offline event. Representation of a project may be further defined and clarified by project maintainers.
+
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting the project team via the [repository maintainers](mailto:eiffel-protocol-maintainers@googlegroups.com). All complaints will be reviewed and investigated and will result in a response that is deemed necessary and appropriate to the circumstances. The project team is obligated to maintain confidentiality with regard to the reporter of an incident. Further details of specific enforcement policies may be posted separately.
```
Comment (fredjn) [inline CODE_OF_CONDUCT.md]: This should apply to all repo's in eiffel-community(?) and not just the protocol. If so then we perhaps need to create another point of contact for reporting abuse etc....

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+* The use of sexualized language or imagery and unwelcome sexual attention or advances
+* Trolling, insulting/derogatory comments, and personal or political attacks
+* Public or private harassment
+* Publishing others' private information, such as a physical or electronic address, without explicit permission
+* Other conduct which could reasonably be considered inappropriate in a professional setting
+
+## Our Responsibilities
+
+Project maintainers are responsible for clarifying the standards of acceptable behavior and are expected to take appropriate and fair corrective action in response to any instances of unacceptable behavior.
```
Comment (fredjn) [inline CODE_OF_CONDUCT.md]: I think this is also the responsibility of the community itself and not just the maintainers, but perhaps that is implied by the pledge above?

File: CONTRIBUTING.md (hunk 1 above)
Comment (fdegir) [reply CONTRIBUTING.md]: Thanks for the review @fredjn. You are right about your comment with regards to contributor guidelines being common across all the projects under Eiffel Community. I missed this while copying it from Eiffel repository.

I'll amend the PR and we can work on it together to formulate the right wording. We need this to be linked from all the projects as well since all of them have the contributor guideline in their repos.

File: CONTRIBUTING.md (hunk 2 above)
Comment (fdegir) [reply CONTRIBUTING.md]: This is something we need to make it more visible - who are the maintainers and how to communicate with them so people can easily find that info.

I'll create a new issue and propose using Github code owners file/feature to increase the visibility of the maintainers and also notify those maintainers for the new PRs.

https://help.github.com/en/articles/about-code-owners

File: CONTRIBUTING.md (hunk 3 above)
Comment (fdegir) [reply CONTRIBUTING.md]: ditto

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (fdegir) [reply CODE_OF_CONDUCT.md]: Again, I missed this while copying it from Eiffel repository.

But this is kind of tricky as well as the practice to report and receive code of conduct issues is to use specific mail address that has very limited access by very few people as the nature of the issue might be sensitive and should not be distributed widely.

This is something we need to take as first item when/if the Technical Committee is formed.

File: CODE_OF_CONDUCT.md (hunk 5 above)
Comment (fdegir) [reply CODE_OF_CONDUCT.md]: We should make it more explicit that the code of conduct guidelines apply to entire community and all the individuals within the community know that they have to abide by the rules. 

As it is common across all the projects, it is more appropriate for Technical Committee to be responsible for it in its entirety (creation, clarification, followup, taking actions, etc) rather than project maintainers. Again, this is another item to have a discussion within TC.

File: CONTRIBUTING.md (hunk 2 above)
Comment (rogpers-ericsson) [reply CONTRIBUTING.md]: Yes, using the code owners file make sense (although I'm not sure I understand all possibilities in using it).

File: CONTRIBUTING.md (hunk 1 above)
Comment (rogpers-ericsson) [reply CONTRIBUTING.md]: Will we risk end up having multiple places for issues, one global one per repo?

File: CODE_OF_CONDUCT.md (hunk 5 above)
Comment (rogpers-ericsson) [reply CODE_OF_CONDUCT.md]: Should it be a responsibility of the owners to raise any issues, rather than doing local things?

File: CONTRIBUTING.md (hunk 1 above)
Comment (fdegir) [reply CONTRIBUTING.md]: I think that will be the case and project specific issues will go to that project's issue tracker/backlog as we don't have a central issue server like Jira or Bugzilla..

The community related issues/backlog items could be submitted to community repo and this repo could be owned by the Technical Committee. TC either needs to take an action and work on it or route it to the right project.

This is yet another item to discuss, clarify, and document.

File: CONTRIBUTING.md (hunk 2 above)
Comment (fdegir) [reply CONTRIBUTING.md]: Some open source projects keep a file in their repositories and list maintainers, meeting logistics, and other project specific information.

Codeowners file seems to be the closest thing to that and helps us not to invent the mechanism ourselves. Other good thing about this file is that Github notifies those people for new PRs so people don't need to send mails etc for getting attention from the maintainers.

"**Code owners are automatically requested for review when someone opens a pull request that modifies code that they own.**"

File: CODE_OF_CONDUCT.md (hunk 5 above)
Comment (fdegir) [reply CODE_OF_CONDUCT.md]: I think as a good citizens, everyone who is a contributor to the community is responsible to raise issues/bring them to attention of TC or the people who are endorsed by TC to work on code of conduct issues, preferably via that private mail alias without publicizing it.

The owners should also follow the same process to raise issues rather than taking action themselves.

TC is responsible to do periodic code of conduct reminders to entire community on a quarterly basis for example rather than just adding the file there and hoping for the best. This will help raise the awareness for the members who joined recently and also make others remember it.

**Comment by henning-roos** [comment]
This PR is obsolete. New eiffel-community/.github repository offers default file for CONTRIBUTING.md

**Comment by fdegir** [comment]
thanks for the comment but i have few questions, comments and perhaps more importantly concerns about this.

i see that the files in the new repo you referred were created by a PR you issued, accepted, and merged yourself instead of working on this open PR or asking others to review your PR on the other repo you created.

I don’t think it should be that way if we want Eiffel to be a real community working on Eiffel in a collaborative manner. 

Apart from that, I don’t see any comment from you on this PR with regards to
- what’s wrong with this PR
- how it could be improved
- why it should be in that repo and not this

The whole idea with this and other related PRs and open issues was to establish clear guidelines and processes for Eiffel that are worked on in a collaborative manner and accepted after community consensus is achieved rather than individuals submitting their own PRs, essentially setting up the processes which community needs to adhere to but without getting community input.

**Comment by henning-roos** [comment]
Thanks for voicing your concerns.

I should have explained myself better.

First, without the basic files required for all eiffel repositories are missing (see eiffel-repository-template), such as CODE_OF_CONDUCT.md and CONTRIBUTING.md, it is pointless 
 to discuss community contributions, in my opinion.

It is better to get them in place first fast and then file issues.

These are duplications of other PR:s filed at the same time (#17, and #18). So I can't merge these without making those in conflict. But those PR:s where also abandoned by the committer (no activity or changes since June, despite comments and requests for changes). I, as a maintainer, is put in a Catch 22 here.

I was invoking number 1 and 3 from the contributing guidelines (https://github.com/eiffel-community/.github/blob/master/CONTRIBUTING.md#closing-issues).

Basically, I try to clean up this repository since no one have tried to do so for a long time.

**Comment by fdegir** [comment]
Thanks for the explanation and it would have been much better to comment this before you invoked those clauses you referred and closed the PR, giving chance to explain the reasons why the PRs are still open and so on.

Here are some more details.

- This and other related PRs might be pointless for you but they are not pointless for quite few people who opened them, reviewed them, and commented on them which you can clearly see in the history. This is simply disregarding people's time and effort, not just my own.
- Secondly, regarding the template repo your refer to: This and other PRs are more about the content and not the cosmetics or the repo structure. The template repo and such will need to be reworked on since all the newly created repos get duplicated documents that should be common for the community and not project specific such as code of conduct, contribution guidelines, and so on.  I looked at that repo before I opened these PRs and filed the issues for the removal of those files so again you disregard other people's time and effort and don't even ask why it is this way.
- I also want to point out that I am not claiming that these PRs are doing the right things or doing things in right way. The only way to find that out is to keep the conversation open rather than closing the PRs singlehandedly.
- I am not expecting these PRs to be accepted or merged as is. It is totally fine for me if they are not accepted as well. The only point with these PRs is to have conversation within **the community** and see if anyone has better ideas or concerns with regards to these as these could become the rules or norms for the community which everyone is expected to adhere to. Also, I think I have merge rights in this repo and I could have invoked my right to merge these PRs but didn't do it both because I didn't address the comments due to various reasons or people didn't review them because of other reasons. One way to find that out why this is the case is just asking what's going on with the PR which could have saved lots of time.

**Comment by m-linner-ericsson** [comment]
I agree with Henning on closing the PR. 
- No one has commented on this PR in 4 months. Could you elaborate on why you believe more people will add to the discussion?
- I find dangling pull request a sign of abandonment. To me, it looks like the requestor forgot the pull request.
- Could you give an example of a repository that keeps its discussion in open pull requests? In my experience, most repositories strive to close the pull request as fast as possible.
- From "Contributing to Eiffel Community": 

> Issues without conclusion and no activity for at least 14 days may be closed, as a mere housekeeping activity. For instance, an Issue met with requests for further clarification, but left unanswered by the original author, may simply be removed.

**Comment by fdegir** [comment]
@m-linner-ericsson as I noted in my comments, my concers were not related to what was done (closure of the PR) but how it was done (closing without asking what's the status of the PR to the contributor).

as you point out, closure of the PRs on Github or changes on Gerrit are common in other open source communities. but it is very different how it is done.
- in case if a PR/change is intended to be closed by humans (maintainers, etc.), they at least show courtesy to the contributor to check with them before closing the PR as the person put effort in that PR.
- if automatic closures are enabled due to inactivity, this is clearly written/communicated that the PR will be closed by the system unless the contributor does not do that.

here are my thoughts on your questions.
- Eiffel summit happened just before the summer and the timing of these PRs were not good as many people were on summer break. After that, I simply didn't have time to follow it up but thad intention to come back. I even mentioned to some maintainers just last week or something verbally. To me, cleanup is good but if it is done in a right way.
- To me dangling PR is not sign up abandonment. The person might have been busy, sick, on holiday or simply not ready with the rest of the work because more discussions are needed. To ensure it is abandoned, the person can be asked about the status of the PR and if no response is received then the PR can be deemed to be abandoned.
- regarding the examples you ask. i can't list all the repos but here are some from other open source communities I'm contributing to.
  - https://review.opendev.org/#/q/owner:%22Fatih+Degirmenci+%253Cfatih.degirmenci%2540ericsson.com%253E%22++status:open
  - https://gerrit.opnfv.org/gerrit/#/q/owner:%22Fatih+Degirmenci+%253Cfdegir%2540gmail.com%253E%22+status:open

  - and an example from a repo on Github where I had a PR which was open for about 5 months and in the end I was notified by the bot to close the PR and what happens if I don't do that. (even the bots have courtesy in open source) as you see there, I went and closed the PR.
    - https://github.com/kubernetes-sigs/kubespray/pull/3923
- what you refer in the contribution guidelines may put this PR into 2nd category i pointed above but to me it is under the first category as it is closed by a maintainer without checking with the contributor. you may see this as a tiny difference but it is more than that as the maintainer didn't even spend time to ask to contributor with just a sentence to check. I'm sure you are all aware that open source is not just about the technologies and so on but about humans. Please think me about someone who found Eiffel by mistake and liked it a lot. Made my first contribution and forgotten about it because I'm doing it just for fun and without any company association or reason. What would I think if my PR gets closed without the community takes time to ask me?

**Comment by m-linner-ericsson** [comment]
Just so I understand you correctly. You would like a similar notification posted on https://github.com/eiffel-community/community/pull/15, but posted on this pull request.

**Comment by fdegir** [comment]
It doesn't have to be exactly the same way I linked because my comments are mainly about how the community works and whether there are things that can be improved. What matters is that the community discusses how they want to work, documents it, and communicates it clearly. 

Some might say that the contribution guidelines clearly states that "Issues without conclusion and no activity for at least 14 days may be closed". But what document doesn't state is what should contributors expect on the 15th day? Will they get some kind of reminder/notification from a bot or will the maintainer who looks at that issue contacts them before removal/closure or? 

What if Eiffel Community thinks the current way works for the community and decides to continue as is; this is is perfectly fine as well but I'm not sure if that would make Eiffel a community we want it to be; a welcoming community who is considerate to its contributors no matter who they are and work for, how much they contribute to Eiffel, how good they are looking after the PRs/issues they opened, and so on.

I suggest 2 things regarding handling stale PRs/issues and notifications; 

- improve the contribution guidelines to clearly state what happens to stale PRs/issues, how they are handled, and what contributors should expect. (a notification by a maintainer or a bot before closure/removal or no notification at all)
- ask the community (via maillist perhaps) to see if they find any value of having a bot to send  reminders for stale PRs/issues and close them automatically if there is no response for X number of days or continue as is handling them manually but with more consideration.

**Comment by m-linner-ericsson** [comment]
I made a proposal to the guidelines concerning the closing of pull request https://github.com/eiffel-community/.github/pull/4.
